### ISSUE TITLE
HPKP security header

### ISSUE THREAD

**Issue by cryptolok** [PrivateBin/PrivateBin#160]
You have well implemented (or almost) security headers for your servers.
One exception in HPKP (HTTP Public Key Pinning) which is not set at all.
I'm just curious to know why.

**Comment by elrido** [comment]
I looked into it, but realized that it requires me to ensure that there is at least one backup key already created up front and stored safely in case I need to replace the key at any point in time. Also I would need to decide and lock in to a time frame for the replacement of the keys (i.e. 2 months is often given as such a period).

This brings a lot more administrative work then just creating a new key on the fly if ever needed (typically before requesting a new certificate or when there is a major security issue with the SSL library that may have caused a leakage of said key).

I see the risk that we could get in a situation were we are locking users out of the service due to not having the organizational structure required to be able to ensure such a second or third key is stored in a safe location (for example it should not be stored at my home on a computer, but offline and offsite, ideally in a safe-box in a bank vault or similar). Although I do personally have access to such, it then requires physical access to replace the key. But if you are in the situation that you want to replace the key, you would rather like to replace it as quick as possible from home and not loose time because you have to wait for the next business day when the bank is open, take a day off at work, commute there, etc. For larger organizations (that do not consist of a just single person running the server) this is certainly manageable. Ideally you'd have a team of 3 - 5 people with access to the backup certs and the server(s) or better yet separate server access and backup key access into two teams of 3 - 5 people each.

Additionally one might even consider the increased risk of having valid backup keys generated up front, in case I would not keep them safe enough and somebody attacking me would gain access to them (if I wouldn't store them offline, for sake of the argument).

Basically I do see only a marginal gain in security over HTTPS with HSTS, but great potential to accidentally shot myself in the foot.

This article summarizes nicely why many sites fear to implemented this:
https://news.netcraft.com/archives/2016/03/22/secure-websites-shun-http-public-key-pinning.html

**Comment by rugk** [comment]
Yes HPKP certainly involves a certain risk. Thankfully you can decrease it by e.g. pinning the root (e.g. Let's encrypt, or multiple ones) and multiple backup keys, so that you can always request a new cert without having to worry too much.
In any case @elrido explained why we currently don't do it on privatebin.net.

And just to complete the thing: of course we cannot set such set/hardcode header in PrivateBin as it must be different by every server, needs some thought and we also cannot access the SSL certs in PrivateBin.

I think this question had been answered in detail, so I am closing this issue.
